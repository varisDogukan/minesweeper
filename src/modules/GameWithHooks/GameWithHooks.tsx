import React from "react";

import { emptyFieldGenerator, CellState } from "@/helpers/Field";
import { Scoreboard } from "@/components/Scoreboard";
import Grid from "@/components/Grid/Grid";
import { GameArea, GameOver, Wrapper } from "@/components/Game";
import Top from "@/components/Top/Top";
import {
  GameLevels,
  LevelNames,
  GameSettings,
  levelNames,
} from "../GameSettings";

export default function GameWithHooks() {
  const [level, setLevel] = React.useState<LevelNames>("beginner");

  const [size, bombs] = GameSettings[level];

  const playerField = emptyFieldGenerator(size, CellState.hidden);

  return (
    <Wrapper>
      <Top feature='Flag' firstAction='ctrl' secondAction='click'>
        Minesweeper
      </Top>
      <GameArea>
        <Scoreboard
          time='0'
          levels={levelNames}
          defaultLevel={level}
          onChangeLevel={({ target: { value } }) => {
            setLevel(value as LevelNames);
          }}
          bombs='10'
          onReset={() => null}
        />
        <GameOver isWin={true} onClick={() => null} />
        <Grid onClick={() => null} onContextMenu={() => null}>
          {playerField}
        </Grid>
      </GameArea>
    </Wrapper>
  );
}
