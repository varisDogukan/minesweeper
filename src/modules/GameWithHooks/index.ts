import GameWithHooks from "./GameWithHooks";

export default GameWithHooks;
