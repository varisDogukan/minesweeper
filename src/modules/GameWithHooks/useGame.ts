import React, { useState } from "react";

import {
  generateFieldWithDefaultState,
  CellState,
  fieldGenerator,
  Coords,
  Field,
} from "@/helpers/Field";
import { LevelNames, GameSettings } from "../GameSettings";
import openCell from "@/helpers/openCell";
import setFlag from "@/helpers/setFlag";

interface ReturnType {
  level: LevelNames;
  isGameOver: boolean;
  isWin: boolean;
  settings: [number, number];
  playerField: Field;
  gameField: Field;
  onClick: (coords: Coords) => void;
  onContextMenu: (coords: Coords) => void;
  onChangeLevel: (level: LevelNames) => void;
  onReset: () => void;
}

export default function useGame(): ReturnType {
  const [level, setLevel] = useState<LevelNames>("beginner");

  const [isGameOver, setIsGameOver] = React.useState(false);
  const [isWin, setIsWin] = React.useState(false);

  const [size, bombs] = GameSettings[level];

  const [playerField, setPlayerField] = useState<Field>(
    generateFieldWithDefaultState(size, CellState.hidden)
  );

  const [gameField, setGameField] = useState<Field>(
    fieldGenerator(size, bombs / (size * size))
  );

  const onClick = (coords: Coords) => {
    try {
      const newPlayerField = openCell(coords, playerField, gameField);
      setPlayerField([...newPlayerField]);
    } catch (error) {
      setPlayerField([...gameField]);
      setIsGameOver(true);
    }
  };

  const onContextMenu = (coords: Coords) => {
    const newPlayerField = setFlag(coords, playerField, gameField);
    setPlayerField([...newPlayerField]);
  };

  const resetHandler = ([size, bombs]: [number, number]) => {
    const newGameField = fieldGenerator(size, bombs / (size * size));
    const newPlayerField = generateFieldWithDefaultState(
      size,
      CellState.hidden
    );

    setGameField([...newGameField]);
    setPlayerField([...newPlayerField]);
    setIsGameOver(false);
    setIsWin(false);
  };

  const onChangeLevel = (level: LevelNames) => {
    setLevel(level);
    const newSettings = GameSettings[level];
    resetHandler(newSettings);
  };

  const onReset = () => resetHandler([size, bombs]);

  return {
    level,
    isGameOver,
    isWin,
    settings: [size, bombs],
    playerField,
    gameField,
    onClick,
    onContextMenu,
    onChangeLevel,
    onReset,
  };
}
