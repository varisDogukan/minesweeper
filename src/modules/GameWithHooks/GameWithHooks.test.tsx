import React from "react";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import GameWithHooks from "./GameWithHooks";

describe("GameWithHooks test cases", () => {
  describe("Render behaviour", () => {
    it("Render game field by default", () => {
      const { asFragment } = render(<GameWithHooks />);

      expect(screen.getAllByRole("cell")).toHaveLength(81);

      expect(asFragment()).toMatchSnapshot();
    });

    it("onChange game level handler", async () => {
      render(<GameWithHooks />);

      expect(screen.getAllByRole("cell")).toHaveLength(81);

      await userEvent.selectOptions(
        screen.getByRole("combobox"),
        "intermediate"
      );
      expect(screen.getAllByRole("cell")).toHaveLength(256);

      await userEvent.selectOptions(screen.getByRole("combobox"), "expert");
      expect(screen.getAllByRole("cell")).toHaveLength(484);
    });
  });
});
