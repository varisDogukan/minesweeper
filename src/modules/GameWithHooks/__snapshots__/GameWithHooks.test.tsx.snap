// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameWithHooks test cases Render behaviour Render game field by default 1`] = `
<DocumentFragment>
  <div
    class="css-1ct4owg"
  >
    <header
      class="css-dn2347"
    >
      <h1
        class="css-18kvaxm"
      >
        Minesweeper
      </h1>
      <legend
        class="css-17qz6a2"
      >
        <strong>
          Flag: 
        </strong>
        <code
          class="css-h9xasx"
        >
          <span
            class="css-udfepp"
          >
            ctrl
          </span>
           + 
          <span
            class="css-9c23k1"
          >
            click
          </span>
        </code>
      </legend>
    </header>
    <div
      class="css-1x1aris"
    >
      <div
        class="css-1u7xlis"
      >
        <div
          class="css-1ksnf8d"
        >
          0
        </div>
        <select
          class="css-1jtcpq4"
        >
          <option
            class="css-zcmdxr"
            value="beginner"
          >
            beginner
          </option>
          <option
            class="css-zcmdxr"
            value="intermediate"
          >
            intermediate
          </option>
          <option
            class="css-zcmdxr"
            value="expert"
          >
            expert
          </option>
        </select>
        <button
          class="css-q9gsot"
        >
          ðŸ˜Œ
        </button>
        <div
          class="css-1ksnf8d"
        >
          10
        </div>
      </div>
      <div
        class="css-1uy3707"
      >
        ðŸ˜Ž
      </div>
      <div
        class="css-1ypakgn"
        size="9"
      >
        <div
          class="css-7e3zdk"
          data-testid="10_0,0"
          role="cell"
        />
        <div
          class="css-7e3zdk"
          data-testid="10_1,0"
          role="cell"
        />
        <div
          class="css-7e3zdk"
          data-testid="10_2,0"
          role="cell"
        />
        <div
          class="css-7e3zdk"
          data-testid="10_3,0"
          role="cell"
        />
        <div
          class="css-7e3zdk"
          data-testid="10_4,0"
          role="cell"
        />
        <div
          class="css-7e3zdk"
          data-testid="10_5,0"
          role="cell"
        />
        <div
          class="css-7e3zdk"
          data-testid="10_6,0"
          role="cell"
        />
        <div
          class="css-7e3zdk"
          data-testid="10_7,0"
          role="cell"
        />
        <div
          class="css-7e3zdk"
          data-testid="10_8,0"
          role="cell"
        />
        <div
          class="css-7e3zdk"
          data-testid="10_0,1"
          role="cell"
        />
        <div
          class="css-7e3zdk"
          data-testid="10_1,1"
          role="cell"
        />
        <div
          class="css-7e3zdk"
          data-testid="10_2,1"
          role="cell"
        />
        <div
          class="css-7e3zdk"
          data-testid="10_3,1"
          role="cell"
        />
        <div
          class="css-7e3zdk"
          data-testid="10_4,1"
          role="cell"
        />
        <div
          class="css-7e3zdk"
          data-testid="10_5,1"
          role="cell"
        />
        <div
          class="css-7e3zdk"
          data-testid="10_6,1"
          role="cell"
        />
        <div
          class="css-7e3zdk"
          data-testid="10_7,1"
          role="cell"
        />
        <div
          class="css-7e3zdk"
          data-testid="10_8,1"
          role="cell"
        />
        <div
          class="css-7e3zdk"
          data-testid="10_0,2"
          role="cell"
        />
        <div
          class="css-7e3zdk"
          data-testid="10_1,2"
          role="cell"
        />
        <div
          class="css-7e3zdk"
          data-testid="10_2,2"
          role="cell"
        />
        <div
          class="css-7e3zdk"
          data-testid="10_3,2"
          role="cell"
        />
        <div
          class="css-7e3zdk"
          data-testid="10_4,2"
          role="cell"
        />
        <div
          class="css-7e3zdk"
          data-testid="10_5,2"
          role="cell"
        />
        <div
          class="css-7e3zdk"
          data-testid="10_6,2"
          role="cell"
        />
        <div
          class="css-7e3zdk"
          data-testid="10_7,2"
          role="cell"
        />
        <div
          class="css-7e3zdk"
          data-testid="10_8,2"
          role="cell"
        />
        <div
          class="css-7e3zdk"
          data-testid="10_0,3"
          role="cell"
        />
        <div
          class="css-7e3zdk"
          data-testid="10_1,3"
          role="cell"
        />
        <div
          class="css-7e3zdk"
          data-testid="10_2,3"
          role="cell"
        />
        <div
          class="css-7e3zdk"
          data-testid="10_3,3"
          role="cell"
        />
        <div
          class="css-7e3zdk"
          data-testid="10_4,3"
          role="cell"
        />
        <div
          class="css-7e3zdk"
          data-testid="10_5,3"
          role="cell"
        />
        <div
          class="css-7e3zdk"
          data-testid="10_6,3"
          role="cell"
        />
        <div
          class="css-7e3zdk"
          data-testid="10_7,3"
          role="cell"
        />
        <div
          class="css-7e3zdk"
          data-testid="10_8,3"
          role="cell"
        />
        <div
          class="css-7e3zdk"
          data-testid="10_0,4"
          role="cell"
        />
        <div
          class="css-7e3zdk"
          data-testid="10_1,4"
          role="cell"
        />
        <div
          class="css-7e3zdk"
          data-testid="10_2,4"
          role="cell"
        />
        <div
          class="css-7e3zdk"
          data-testid="10_3,4"
          role="cell"
        />
        <div
          class="css-7e3zdk"
          data-testid="10_4,4"
          role="cell"
        />
        <div
          class="css-7e3zdk"
          data-testid="10_5,4"
          role="cell"
        />
        <div
          class="css-7e3zdk"
          data-testid="10_6,4"
          role="cell"
        />
        <div
          class="css-7e3zdk"
          data-testid="10_7,4"
          role="cell"
        />
        <div
          class="css-7e3zdk"
          data-testid="10_8,4"
          role="cell"
        />
        <div
          class="css-7e3zdk"
          data-testid="10_0,5"
          role="cell"
        />
        <div
          class="css-7e3zdk"
          data-testid="10_1,5"
          role="cell"
        />
        <div
          class="css-7e3zdk"
          data-testid="10_2,5"
          role="cell"
        />
        <div
          class="css-7e3zdk"
          data-testid="10_3,5"
          role="cell"
        />
        <div
          class="css-7e3zdk"
          data-testid="10_4,5"
          role="cell"
        />
        <div
          class="css-7e3zdk"
          data-testid="10_5,5"
          role="cell"
        />
        <div
          class="css-7e3zdk"
          data-testid="10_6,5"
          role="cell"
        />
        <div
          class="css-7e3zdk"
          data-testid="10_7,5"
          role="cell"
        />
        <div
          class="css-7e3zdk"
          data-testid="10_8,5"
          role="cell"
        />
        <div
          class="css-7e3zdk"
          data-testid="10_0,6"
          role="cell"
        />
        <div
          class="css-7e3zdk"
          data-testid="10_1,6"
          role="cell"
        />
        <div
          class="css-7e3zdk"
          data-testid="10_2,6"
          role="cell"
        />
        <div
          class="css-7e3zdk"
          data-testid="10_3,6"
          role="cell"
        />
        <div
          class="css-7e3zdk"
          data-testid="10_4,6"
          role="cell"
        />
        <div
          class="css-7e3zdk"
          data-testid="10_5,6"
          role="cell"
        />
        <div
          class="css-7e3zdk"
          data-testid="10_6,6"
          role="cell"
        />
        <div
          class="css-7e3zdk"
          data-testid="10_7,6"
          role="cell"
        />
        <div
          class="css-7e3zdk"
          data-testid="10_8,6"
          role="cell"
        />
        <div
          class="css-7e3zdk"
          data-testid="10_0,7"
          role="cell"
        />
        <div
          class="css-7e3zdk"
          data-testid="10_1,7"
          role="cell"
        />
        <div
          class="css-7e3zdk"
          data-testid="10_2,7"
          role="cell"
        />
        <div
          class="css-7e3zdk"
          data-testid="10_3,7"
          role="cell"
        />
        <div
          class="css-7e3zdk"
          data-testid="10_4,7"
          role="cell"
        />
        <div
          class="css-7e3zdk"
          data-testid="10_5,7"
          role="cell"
        />
        <div
          class="css-7e3zdk"
          data-testid="10_6,7"
          role="cell"
        />
        <div
          class="css-7e3zdk"
          data-testid="10_7,7"
          role="cell"
        />
        <div
          class="css-7e3zdk"
          data-testid="10_8,7"
          role="cell"
        />
        <div
          class="css-7e3zdk"
          data-testid="10_0,8"
          role="cell"
        />
        <div
          class="css-7e3zdk"
          data-testid="10_1,8"
          role="cell"
        />
        <div
          class="css-7e3zdk"
          data-testid="10_2,8"
          role="cell"
        />
        <div
          class="css-7e3zdk"
          data-testid="10_3,8"
          role="cell"
        />
        <div
          class="css-7e3zdk"
          data-testid="10_4,8"
          role="cell"
        />
        <div
          class="css-7e3zdk"
          data-testid="10_5,8"
          role="cell"
        />
        <div
          class="css-7e3zdk"
          data-testid="10_6,8"
          role="cell"
        />
        <div
          class="css-7e3zdk"
          data-testid="10_7,8"
          role="cell"
        />
        <div
          class="css-7e3zdk"
          data-testid="10_8,8"
          role="cell"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
